### PR TITLE
TPC: automatizing creation of IDC scalers during synchronous reco

### PR DESCRIPTION
**Pull request by matthias-kleiner**

### CODE REVIEW

**Comment by matthias-kleiner** [comment]
Hello @shahor02 ,

in my local tests it happened sometimes that two or sometimes even three identical objects were written to the test ccdb http://ccdb-test.cern.ch:8080/browse/TPC/Calib/Scaler. For example for the timestamp 1717055684611:
test_ccdb_identical
Do you know how this can happen? Could this be as the objects are `TTrees`? When I was storing the object directly I didnt observe it...

**Comment by shahor02** [comment]
@matthias-kleiner the DPL and the ccdb-populator are agnostic on the object content. 2 objects mean that they were indeed sent twice. You have lot of logs in your PR, what do they tell?

File: Detectors/TPC/workflow/include/TPCWorkflow/TPCFourierTransformAggregatorSpec.h [hunk 1]
```diff
@@ -179,19 +244,165 @@ class TPCFourierTransformAggregatorSpec : public o2::framework::Task
...
   }
+
+  void makeTPCScaler(DataAllocator& output, const bool eos)
+  {
+    LOGP(info, "Making TPC scalers");
+    // check if IDC scalers can be created - check length of continous received IDCs
+    std::vector<std::pair<long, long>> times;
+    times.reserve(mTPCScalerCont.idcs.size());
+    for (const auto& idc : mTPCScalerCont.idcs) {
+      times.emplace_back(idc.second.timestamp[0], idc.second.timestamp[1]);
```
Comment (shahor02) [inline Detectors/TPC/workflow/include/TPCWorkflow/TPCFourierTransformAggregatorSpec.h:255]: btw, your mTPCScalerCont is an unordered_map, in looping in this way, don't you risk to upload not in time ordered way? I think std::map would be safer.

Comment (matthias-kleiner) [reply Detectors/TPC/workflow/include/TPCWorkflow/TPCFourierTransformAggregatorSpec.h:255]: thats why I fill the time information to the vector and sort this vector afterward and then loop over the sorted vector.

**Comment by matthias-kleiner** [comment]
> @matthias-kleiner the DPL and the ccdb-populator are agnostic on the object content. 2 objects mean that they were indeed sent twice. You have lot of logs in your PR, what do they tell?

Yes, there are some logs. I thought it would be good to have something there in case we will get problems in the online processing. As the run function gets triggered only every 11 seconds, the logs are not printed very often.

**Comment by shahor02** [comment]
But should not  the log tell you if you indeed have 2 slots with the same timespan or a single slot is sent twice?

**Comment by matthias-kleiner** [comment]
> But should not the log tell you if you indeed have 2 slots with the same timespan or a single slot is sent twice?

In my logs I didnt see that the same object was sent twice. It looked like that the problem is in the ccdb-populator, but I can re-run the test and check again

**Comment by shahor02** [comment]
Please do, the ccdb-populator uploads what it receives, we have never seen different behaviour.

**Comment by matthias-kleiner** [comment]
> Please do, the ccdb-populator uploads what it receives, we have never seen different behaviour.

I was just running it again. There are again 2 objects stored, which are identical. In the log I see  from the ccdb-populator only once that its stored:
`[1527708:ccdb-populator]: [16:57:33][INFO] Storing in ccdb TPC/Calib/Scaler/TTree_1717081053459.root of size 2097152 valid for 1717080181227 : 1717080215374
`
grafik

The log file:
[log.txt](https://github.com/AliceO2Group/AliceO2/files/15502352/log.txt)

**Comment by shahor02** [comment]
Thanks @matthias-kleiner , then it should be a server side or CURL problem.

@costing, the ccdb-populator uploads as many times as it reports in the log: https://github.com/AliceO2Group/AliceO2/blob/3a39b254899ba709cef28d5535f6ede48f213657/Detectors/Calibration/workflow/CCDBPopulatorSpec.h#L104-L120.
But for a single report 
```
[1527708:ccdb-populator]: [16:57:33][INFO] Storing in ccdb TPC/Calib/Scaler/TTree_1717081053459.root of size 2097152 valid for 1717080181227 : 1717080215374
```
there are 2 objects stored with the difference of 0.5s: http://ccdb-test.cern.ch:8080/browse/TPC/Calib/Scaler/1717080181227.
And this happens only once in a while. 
Any idea what could be the reason?

**Comment by costing** [comment]
Hi @shahor02 ,

On the server side I see two independent POST requests, they arrive at 2 s difference but the first one took >1 s to be written so in the end the recorded difference is just 0.5 s . 
```
141.2.243.5 - - [30/May/2024:16:57:33 +0200] "POST /TPC/Calib/Scaler/1717080181227/1717080215374/ObjectType=TTree/adjustableEOV=true/ HTTP/1.1" 201 714 "-" "pcikf5-1717079917-KO64mj"
141.2.243.5 - - [30/May/2024:16:57:35 +0200] "POST /TPC/Calib/Scaler/1717080181227/1717080215374/ObjectType=TTree/adjustableEOV=true/ HTTP/1.1" 201 714 "-" "pcikf5-1717079917-KO64mj"
```
So it looks like the client performs it twice, but it's weird that you only see one request logged.

Cheers,

.costin

**Comment by shahor02** [comment]
thanks for checking, @costing , can it be curl problem? The populated calls only once the ccdbAPI::storeAsBinaryFile.
This latter has this loop: https://github.com/AliceO2Group/AliceO2/blob/3a39b254899ba709cef28d5535f6ede48f213657/CCDB/src/CcdbApi.cxx#L410-L427, but since duplication happened only once, I would exclude that the `hostsPool.size()` could be >1.

**Comment by costing** [comment]
Hi @shahor02 ,

It might be something else in this logic. Can it be that the timeout is set to 1s via this code path? https://github.com/AliceO2Group/AliceO2/blob/3a39b254899ba709cef28d5535f6ede48f213657/CCDB/src/CcdbApi.cxx#L234-L235
or from outside with `$ALICEO2_CCDB_CURL_TIMEOUT` ?

Then the `CURLE_OPERATION_TIMEDOUT` condition would be triggered and the upload would be retried.

For read requests we can have small timeouts as they are redirects to AliEn, but uploads can take significantly longer and are a function of the object size. So the `storeAsBinaryFile` method would benefit from a different request timeout value than the `receive*` ones.

Cheers,

.costin

**Comment by shahor02** [comment]
Thanks @costing , the CURLE_OPERATION_TIMEDOUT was not set but indeed, in @matthias-kleiner test it should have been set to 1 s via https://github.com/AliceO2Group/AliceO2/blob/3a39b254899ba709cef28d5535f6ede48f213657/CCDB/src/CcdbApi.cxx#L235.

I can introduce a separate upload timeout, what value should I use? Just to be sure how it works: should not it to abort previous upload at the time-out, before trying again?

**Comment by costing** [comment]
Hi @shahor02 ,

I think it depends of the state it was in. Looks like the server had all the data in memory and was in the process of committing it and just the confirmation message was missed.

Can we try a value of `(5 + objectSize / 1MB) seconds` ?

Cheers,

.costin

**Comment by matthias-kleiner** [comment]
Hello @shahor02 
can this PR be merged? The error on macOS seems unrelated.

**Pull request merged**